node/src/Pos/Block/Network/Logic.hs

```diff
@@ -309,10 +310,19 @@ handleRequestedHeaders
     -> NewestFirst NE (BlockHeader ssc)
     -> m (Maybe t)
 handleRequestedHeaders cont inRecovery headers = do
-    classificationRes <- classifyHeaders inRecovery headers
-    let newestHeader = headers ^. _Wrapped . _neHead
+    let newestHeader = headers ^. _NewestFirst . _neHead
+        oldestHeader = headers ^. _NewestFirst . _neLast
         newestHash = headerHash newestHeader
-        oldestHash = headerHash $ headers ^. _Wrapped . _neLast
+        oldestHash = headerHash oldestHeader
+        oldestEpoch = oldestHeader ^. epochIndexL
+    -- Try to calculate LRC for the oldest header epoch. If we're in
+    -- recovery and oldest header is from the next epoch, no lrc will
+    -- be automatically calculated as all workers are locked in
+    -- recovery mode. So we should try to do it manually.
+    logDebug "handleREquesteHeaders: LRC started"
+    void $ tryAny $ withStateLockNoMetrics LowPriority $ const $ lrcSingleShot oldestEpoch
```

Comment:
So where is exception handling? Apparently you just deleted the comment.
1. I believe it's ok to ignore `UnknownBlocksForLrc` here, but it's the only exception we can ignore here. All other exceptions should be at least logged (or maybe not caught at all).
2. Everytime you use `tryAny` or any other function which catches `ErrorCall` or `CardanoFatalError` you should use `reportOrLog` inside exception handler.

Reply:
And, to be honest, I would change this code to explicitly check our tip's slot and call LRC only when we know it will succeed (i. e. our tip's slot is from `oldestEpoch - 1` and it's after crucial slot). Also it makes sense to check chain quality, because see CSL-1660.
This way you don't need to take locks (and note that it will take some time to release it, because LRC loads up to `2k` blocks from the DB) and use exceptions for control flow.

Reply:
@gromakovsky I've implement conditional calculation, though i didn't understand what did you mean by "check chain quality explicitly". I will amend if you decide that it's still needed.

Reply:
@volhovM the thing is that now you still can get `UnknownBlocksForLrc` because LRC checks that there is at least one block after `crucialSlot`. So if tip is from slot `(e, 0)` and you receive a header from slot `(e + 1, 0)`, LRC will fail. So you should check that there is at least one block after crucial slot.
And according to CSL-1660, you should do a stricter check that there are at least `k` blocks. This check can be done without loading blocks, because we store slots of last `k` blocks. But I don't mind if you check presence of only 1 block and add a TODO with `CSL-1660` tag (also write about it in CSL-1660).

Reply:
@gromakovsky please check it again, i've added it.